client/components/Editor/schemas/reactive/structuredCitation.js

```diff
@@ -0,0 +1,24 @@
+/* eslint-disable react-hooks/rules-of-hooks */
+
+export const structuredCitation = (structuredAttr) => {
+	return function(node) {
+		const { [structuredAttr]: value, customLabel } = node.attrs;
+		const { useEffect, useState, useDocumentState } = this;
+		const { citationManager } = useDocumentState();
+		const shouldDeriveCitation = citationManager && !customLabel;
+
+		const [citation, setCitation] = useState(
+			() => shouldDeriveCitation && citationManager.getSync(value),
+		);
+
+		useEffect(() => {
+			if (shouldDeriveCitation) {
+				const unsubscribe = citationManager.subscribe(value, setCitation);
+				return unsubscribe;
+			}
+			return () => {};
+		}, [citationManager, customLabel, shouldDeriveCitation, value]);
+
+		return citation;
```

Comment:
I need to dig in a bit more, but out of curiosity, how often do these callback functions execute? I wonder if we could ditch the React-style hooks with a similar approach that Vue takes if we feel that the former incurs a significant performance hit.

The syntax is super similar, but to oversimplify it, the outer function would execute **only** once.

For example, the `useEffect` hook would function as a true subscriber because would register a callback to be executed each update instead of executing a fresh callback when an array of dependencies changes.

There's a lot more to discuss here but in essence we'd use a similar API that might incur less garbage and avoid the stale closure problem.

This site is a good reference for what I'm describing:

https://composition-api.vuejs.org/#basic-example

Reply:
(Thinking out loud here since I don't have a definitive idea about performance or other API shapes at the moment)

The callbacks — the outer `(node: Node) => T` ones, not the `useEffect` ones — execute every time the document contents change. There's no getting around that altogether, since while some reactive attributes are purely a function of the node's non-reactive attributes, others (like the `counter` one) could have their value invalidated at any time by changes beyond the scope of a single node. We could do some optimizations, like marking some reactive attrs as pure functions of stored attrs, or (my personal favorite idea) just debouncing all of the reactive calculations if they start happening too quickly, since they no longer need to happen synchronously to maintain document integrity. But it's definitely got the same kind of overhead tradeoff as writing React components with hooks, and my goal is for reactive nodes to weigh in roughly as small React components performance-wise, in the sense that adding a few hundred or even a thousand of them on the page will affect, but not absolutely tank, performance.

So, as I was just saying on Slack — it would be awesome if we could get the per-attribute render overhead down, and let's sketch some alternatives and see how appealing they are...I will say that the linked Vue RFC presents a _very_ aesthetically pleasing API, and didn't make me go "blegh" like I did when I first saw React Hooks. But the #1 editor performance bottleneck by a country mile is the React coupling — which is driven by the number of `editorChangeObjects` emitted from the editor, which is driven by the number of Prosemirror transactions — and I want to frame editor performance discussions around that, in the sense that if we have more time to hone performance, there's a wide scope of potentially higher-yield problems to look at.